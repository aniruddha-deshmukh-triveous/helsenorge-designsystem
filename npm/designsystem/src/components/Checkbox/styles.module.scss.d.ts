export type Styles = {
  checkbox: string;
  checkbox__icon: string;
  'checkbox__icon-wrapper': string;
  'checkbox__icon-wrapper--checked': string;
  'checkbox__icon-wrapper--disabled': string;
  'checkbox__icon-wrapper--invalid': string;
  'checkbox__icon-wrapper--on-blueberry': string;
  'checkbox__icon-wrapper--on-dark': string;
  'checkbox-wrapper': string;
  'checkbox-wrapper--bigform': string;
  'checkbox-wrapper--disabled': string;
  'checkbox-wrapper--invalid': string;
  'checkbox-wrapper--on-blueberry': string;
  'checkbox-wrapper--on-dark': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
