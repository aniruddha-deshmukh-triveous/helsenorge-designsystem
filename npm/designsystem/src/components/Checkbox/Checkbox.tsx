import React, { useState } from 'react';

import classNames from 'classnames';

import Check from '../Icons/Check';

import checkboxStyles from './styles.module.scss';

interface CheckboxProps {
  /** Adds custom classes to the element. */
  className?: string;
  /** Om checkbox er checked */
  checked?: boolean;
  /** Teksten som st√•r ved siden av checkbox */
  label: string;
}

export const Checkbox = React.forwardRef((props: CheckboxProps, ref: React.Ref<HTMLUListElement>) => {
  const { className = '', label = '' } = props;
  const [checked, setChecked] = useState(props.checked);

  const checkboxClasses = classNames(checkboxStyles.checkbox, className);

  return (
    <label className={checkboxStyles['checkbox-wrapper']}>
      <input
        className={checkboxClasses}
        type={'checkbox'}
        checked={checked}
        onClick={() => {
          setChecked(!checked);
        }}
      />
      {label}
    </label>
  );
});

export default Checkbox;
