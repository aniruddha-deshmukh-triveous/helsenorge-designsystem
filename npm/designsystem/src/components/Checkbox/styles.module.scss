@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';
@import '../../scss/font-settings.scss';

.checkbox-errors {
  font-size: $font-size-sm;
  font-weight: 600;
  color: $cherry500;
  margin-top: 0;
}

.checkbox-wrapper {
  margin: getSpacer(s) 0;

  &--with-error {
    padding: getSpacer(s) getSpacer(2xs) getSpacer(m);
    background-color: $cherry100;
    border-left: 0.25rem solid $cherry500;

    @media (min-width: map-get($grid-breakpoints, md)) {
      padding: getSpacer(s) getSpacer(m) getSpacer(l);
    }
  }

  &--with-error#{&}--bigform {
    padding: getSpacer(m) getSpacer(m);
  }

  &--bigform {
    margin: getSpacer(xs) 0 getSpacer(xs) 0;
  }
}

.checkbox-wrapper:first-of-type {
  margin-top: 0;
}

.checkbox-wrapper:last-of-type {
  margin-bottom: 0;
}

.checkbox-label {
  display: flex;
  align-items: center;
  cursor: pointer;
  font-size: $font-size-sm;

  @media (min-width: map-get($grid-breakpoints, md)) {
    font-size: $font-size-md;
  }

  &--on-dark {
    color: $white;
  }

  &--invalid {
    color: $black;
  }

  &--disabled {
    cursor: default;
    color: $neutral400;
  }

  &--bigform {
    padding: getSpacer(s);
    background-color: $neutral50;
  }

  &--bigform#{&}--on-blueberry {
    background-color: $blueberry50;
  }

  &--bigform#{&}--on-dark {
    background-color: $blueberry800;
  }

  &--bigform#{&}--invalid {
    background-color: $white;
  }

  @media (min-width: map-get($grid-breakpoints, md)) {
    &--bigform {
      padding: 1.68rem getSpacer(l);
    }
  }
}

.checkbox {
  width: 0px;
  margin: 0;

  -moz-appearance: none;
  -webkit-appearance: none;
  -o-appearance: none;

  &__icon-wrapper {
    display: flex;
    justify-content: center;
    align-items: center;
    margin: getSpacer(4xs) getSpacer(s) getSpacer(4xs) getSpacer(4xs);
    height: 1.625rem;
    width: 1.625rem;
    min-width: auto !important;
    min-height: auto !important;
    color: $neutral700;
    box-shadow: 0 0 0 getSpacer(4xs);

    :hover > & {
      box-shadow: 0 0 0 getSpacer(3xs);
      background-color: $neutral100;
    }

    :active > & {
      box-shadow: 0 0 0 getSpacer(3xs);
      color: $black;
    }

    &--checked {
      color: $blueberry500;

      :hover > & {
        background-color: $blueberry50;
      }
    }

    &--on-blueberry {
      color: $blueberry500;

      :hover > & {
        background-color: $blueberry100;
      }
    }

    &--on-dark {
      color: $neutral200;

      :hover > & {
        background-color: $inverted-hover;
      }

      &.checkbox__icon-wrapper--checked {
        color: $blueberry200;
      }

      :active > & {
        color: $white;
      }
    }

    &--invalid {
      color: $cherry500;

      :hover > & {
        background-color: $cherry100;
      }

      &.checkbox__icon-wrapper--checked {
        color: $cherry500;
      }

      :active > & {
        color: $black;
      }
    }

    &--disabled {
      color: $neutral400;
      &.checkbox__icon-wrapper--checked {
        color: $neutral400;
      }
      :hover > & {
        box-shadow: 0 0 0 getSpacer(4xs);
        background-color: transparent;
      }

      :active > & {
        color: $neutral400;
      }
    }
  }

  &__icon {
    position: relative;
    top: 0.1rem;
    right: 0.05rem;
  }

  &:focus {
    & + .checkbox__icon-wrapper {
      box-shadow: 0 0 0 getSpacer(3xs);
      color: $black;

      &--on-dark {
        color: $white;
      }

      &--invalid {
        color: $black;
      }
    }
  }
}
