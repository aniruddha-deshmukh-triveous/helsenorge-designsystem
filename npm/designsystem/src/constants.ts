export enum IconSize {
  XSmall = 38,
  Small = 48,
  Medium = 64,
  Large = 80,
  XLarge = 130,
}

export enum ZIndex {
  Modal = 1300000,
}

export interface HTMLButtonProps {
  autoFocus?: boolean;
  disabled?: boolean;
  form?: string;
  formAction?: string;
  formEncType?: string;
  formMethod?: string;
  formNoValidate?: boolean;
  formTarget?: string;
  name?: string;
  type?: 'submit' | 'reset' | 'button' | string;
  value?: string | string[] | number;
}

export interface HTMLTextareaProps {
  autoFocus?: boolean;
  disabled?: boolean;
  form?: string;
  formNoValidate?: boolean;
  formTarget?: string;
  name?: string;
  id?: string;
  autocomplete?: 'off' | 'on';
  placeholder?: string;
  readonly?: boolean;
  required?: boolean;
  spellcheck?: 'true' | 'default' | 'false';
  'aria-invalid'?: boolean;
}

export interface HTMLAnchorProps {
  download?: string;
  href?: string;
  hrefLang?: string;
  media?: string;
  ping?: string;
  rel?: string;
  target?: string;
  referrerPolicy?:
    | ''
    | 'same-origin'
    | 'no-referrer'
    | 'no-referrer-when-downgrade'
    | 'origin'
    | 'origin-when-cross-origin'
    | 'strict-origin'
    | 'strict-origin-when-cross-origin'
    | 'unsafe-url';
}

export type ButtonVariant = 'secondary' | 'tertiary' | string | undefined | null;

export enum FormMode {
  onwhite = 'onwhite',
  ongrey = 'ongrey',
  onblueberry = 'onblueberry',
  ondark = 'ondark',
  oninvalid = 'oninvalid',
}

export enum FormVariant {
  normal = 'normal',
  bigform = 'bigform',
}
