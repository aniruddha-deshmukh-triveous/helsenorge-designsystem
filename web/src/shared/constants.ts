export const HNDS_BUCKET_IMAGE_URL = ''; // Kan endres her hvis bilder skal hostes et annet sted

export interface StyledDefaultProps {
  children?: React.ReactNode;
  className?: string;
}
