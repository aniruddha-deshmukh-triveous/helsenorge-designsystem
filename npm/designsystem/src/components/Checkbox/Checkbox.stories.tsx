import React from 'react';
import { boolean, select, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { FormMode, FormVariant } from '../../constants';
import Checkbox from './Checkbox';
import { mapToBackgoundColor } from '../../../.storybook/StoryBackground';

const stories = storiesOf('Checkbox', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => {
  const modes = select('background', FormMode, FormMode.onwhite);

  return (
    <div
      style={{
        width: '40rem',
        background: mapToBackgoundColor(modes),
        padding: '1rem',
      }}
    >
      <Checkbox
        label={'Jeg er frisk!'}
        checked={boolean('Checked', false)}
        disabled={boolean('Disabled', false)}
        mode={modes}
        variant={select('Variant', FormVariant, FormVariant.normal)}
      />
    </div>
  );
});
