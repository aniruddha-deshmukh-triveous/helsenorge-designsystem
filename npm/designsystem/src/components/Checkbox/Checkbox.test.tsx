import React from 'react';

import { render, screen, fireEvent } from '@testing-library/react';

import Checkbox from './Checkbox';
import { FormMode, FormVariant } from '../../constants';

describe('Gitt at Checkbox skal vises', (): void => {
  describe('Når Checkbox rendres', (): void => {
    test('Så vises Checkbox', (): void => {
      const { container } = render(<Checkbox inputId={'test01'} label={'Check me out!'} />);

      expect(container).toMatchSnapshot();

      const label = screen.getByText('Check me out!').parentElement;
      expect(label).toBeVisible();
      expect(label.className).toBe('checkbox-label');

      const input = screen.getByRole('checkbox');
      expect(input).toBeVisible();
      expect(input.className).toBe('checkbox');

      const checkIconWrapper = label.children[1];
      expect(checkIconWrapper.className).toBe('checkbox__icon-wrapper');
    });
  });

  describe('Når disabled er true', (): void => {
    test('Så vises Checkbox som disabled', (): void => {
      render(<Checkbox label={'Check me out!'} disabled />);

      const label = screen.getByText('Check me out!').parentElement;
      expect(label.className).toBe('checkbox-label checkbox-label--disabled');
    });
  });

  describe('Når mode er onBlueberry', (): void => {
    test('Så vises Checkbox med onBlueberry styling', (): void => {
      render(<Checkbox label={'Check me out!'} mode={FormMode.onblueberry} />);

      const label = screen.getByText('Check me out!').parentElement;
      expect(label.className).toBe('checkbox-label');
    });
  });
  describe('Når mode er onDark', (): void => {
    test('Så vises Checkbox med onDark styling', (): void => {
      render(<Checkbox label={'Check me out!'} mode={FormMode.ondark} />);

      const label = screen.getByText('Check me out!').parentElement;
      expect(label.className).toBe('checkbox-label checkbox-label--on-dark');
    });
  });

  describe('Når formvariant er bigform', (): void => {
    test('Så vises Checkbox med bigform styling', (): void => {
      render(<Checkbox label={'Check me out!'} variant={FormVariant.bigform} />);

      const label = screen.getByText('Check me out!').parentElement;
      expect(label.className).toBe('checkbox-label checkbox-label--bigform');
    });
  });

  describe('Når startChecked er true', (): void => {
    test('Så vises Checkbox med checkmark ikon', (): void => {
      render(<Checkbox label={'Check me out!'} checked />);

      const checkIcon = screen.getByText('Check me out!').parentElement.children[1];

      expect(checkIcon).toBeVisible();
      expect(checkIcon.className).toBe('checkbox__icon-wrapper checkbox__icon-wrapper__regular--checked');
    });
  });

  describe('Når checkbox trykkes på', (): void => {
    test('Så vises Checkbox med checkmark ikon', (): void => {
      render(<Checkbox label={'Check me out!'} />);

      const label = screen.getByText('Check me out!').parentElement;

      fireEvent.click(screen.getByText('Check me out!'));

      const checkIcon = label.children[1];

      expect(checkIcon.className).toBe('checkbox__icon-wrapper checkbox__icon-wrapper__regular--checked');

      fireEvent.click(screen.getByText('Check me out!'));

      expect(checkIcon.className).toBe('checkbox__icon-wrapper');
    });
  });

  describe('Når checkbox får satt error', (): void => {
    test('Så vises Checkbox med indre error styling, uten ytre error styling', (): void => {
      render(<Checkbox label={'Check me out!'} error />);

      // Indre styling
      const label = screen.getByText('Check me out!').parentElement;
      const checkIcon = label.children[1];

      expect(label.className).toBe('checkbox-label');
      expect(checkIcon.className).toBe('checkbox__icon-wrapper checkbox__icon-wrapper--on-invalid checkbox__icon-wrapper--invalid');

      // Ytre styling
      const wrapper = label.parentElement;
      expect(wrapper.className).toBe('checkbox-wrapper');
    });
  });

  describe('Når checkbox får satt errorText', (): void => {
    test('Så vises Checkbox med errormelding i tilleg til indre og ytre error styling', (): void => {
      render(<Checkbox label={'Check me out!'} errorText={'error error!'} />);

      expect(screen.getByText('error error!')).toBeVisible();

      // Indre styling
      const label = screen.getByText('Check me out!').parentElement;
      const checkIcon = label.children[1];

      expect(label.className).toBe('checkbox-label');
      expect(checkIcon.className).toBe('checkbox__icon-wrapper checkbox__icon-wrapper--on-invalid checkbox__icon-wrapper--invalid');

      // Ytre styling
      const wrapper = label.parentElement;
      expect(wrapper.className).toBe('checkbox-wrapper checkbox-wrapper--with-error');
    });
  });

  describe('Når name-prop er satt', (): void => {
    test('Så har input riktig name', (): void => {
      render(<Checkbox label="En fin label" name="custom-name" />);

      const checkbox = screen.getByLabelText('En fin label');
      expect(checkbox).toHaveAttribute('name', 'custom-name');
    });
  });

  describe('Når value-prop er satt', (): void => {
    test('Så har input riktig value', (): void => {
      render(<Checkbox label="En fin label" value="custom-value" />);

      const checkbox = screen.getByLabelText('En fin label');
      expect(checkbox).toHaveAttribute('value', 'custom-value');
    });
  });

  describe('Når required er satt', (): void => {
    test('Så er input required', (): void => {
      render(<Checkbox label="En fin label" required />);

      const checkbox = screen.getByLabelText('En fin label');
      expect(checkbox).toBeRequired();
    });
  });

  describe('Når disabled er satt', (): void => {
    test('Så er input disabled', (): void => {
      render(<Checkbox label="En fin label" disabled />);

      const checkbox = screen.getByLabelText('En fin label');
      expect(checkbox).toBeDisabled();
    });
  });

  describe('Når checked er satt', (): void => {
    test('Så er input checked', (): void => {
      render(<Checkbox label="En fin label" checked />);

      const checkbox = screen.getByLabelText('En fin label');
      expect(checkbox).toHaveAttribute('checked', '');
    });
  });
});
