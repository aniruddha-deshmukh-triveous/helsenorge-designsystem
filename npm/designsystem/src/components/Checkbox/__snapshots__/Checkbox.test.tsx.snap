// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Checkbox skal vises Når Checkbox rendres Så vises Checkbox 1`] = `
<div>
  <div
    class="checkbox-wrapper"
    data-analyticsid="checkbox"
  >
    <label
      class="checkbox-label"
      for="test01"
    >
      <input
        aria-invalid="false"
        class="checkbox"
        id="test01"
        name="test01"
        type="checkbox"
        value="Check me out!"
      />
      <span
        class="checkbox__icon-wrapper"
      />
      Check me out!
    </label>
  </div>
</div>
`;
